Maple2.Server.Game/Manager/ItemBoxManager.cs

```diff
@@ -29,11 +31,120 @@ public class ItemBoxManager {
             ItemFunction.SelectItemBox => SelectItemBox(item, itemBoxParams[0], itemBoxParams[1], index, count),
             ItemFunction.OpenItemBox => OpenItemBox(item, itemBoxParams[0], itemBoxParams[1], itemBoxParams[2], itemBoxParams[3], itemBoxParams.Length == 5 ? itemBoxParams[4] : 1, count),
             ItemFunction.OpenItemBoxWithKey => OpenItemBoxWithKey(item, itemBoxParams[0], itemBoxParams[1], itemBoxParams[2], itemBoxParams[5], count),
+            ItemFunction.OpenGachaBox => OpenGachaBox(item, itemBoxParams[0], itemBoxParams[1], count),
             _ => throw new ArgumentOutOfRangeException(item.Metadata.Function?.Type.ToString(), "Invalid box type"),
         };
     }
 
-    private ItemBoxError SelectItemBox(Item item, int itemRequiredAmount, int boxId, int index, int count = 1) { 
+    public ItemBoxError OpenLulluBox(Item item, int count = 1, bool autoPay = false) {
+        Dictionary<string, string> parameters = XmlParseUtil.GetParameters(item.Metadata.Function?.Parameters);
+
+        // Get common dropbox
+        if (!int.TryParse(parameters["commonBoxId"], out int commonBoxId) ||
+            !session.TableMetadata.IndividualItemDropTable.Entries.TryGetValue(commonBoxId, out Dictionary<byte, IList<IndividualItemDropTable.Entry>>? commonDropGroupTable)) {
+            return ItemBoxError.s_err_cannot_open_multi_itembox_inventory_fail;
+        }
+
+        // Get uncommon dropbox
+        if (!int.TryParse(parameters["unCommonBoxId"], out int unCommonBoxId) ||
+            !session.TableMetadata.IndividualItemDropTable.Entries.TryGetValue(unCommonBoxId, out Dictionary<byte, IList<IndividualItemDropTable.Entry>>? unCommonDropGroupTable)) {
+            return ItemBoxError.s_err_cannot_open_multi_itembox_inventory_fail;
+        }
+
+        var boxIngredient = new IngredientInfo(Enum.Parse<ItemTag>(parameters["boxItemTag"]), 1);
+        var keyIngredient = new IngredientInfo(Enum.Parse<ItemTag>(parameters["keyItemTag"]), 1);
+        var totalItems = new List<Item>();
+        for (int startCount = 0; startCount < count; startCount++) {
+            if (!session.Item.Inventory.Consume(new[] {keyIngredient})) {
+                return ItemBoxError.s_err_cannot_open_multi_itembox_inventory;
+            }
+
+            if (autoPay) {
+                int.TryParse(parameters["boxPrice"], out int mesoCost);
+                if (!session.Item.Inventory.Consume(new[] {boxIngredient})
+                    && mesoCost > 0 && session.Currency.CanAddMeso(-mesoCost) == -mesoCost) {
```

Comment:
mesos should be checked before consume is called because consume will mutate the inventory.

Might be simpler to change it to something like:
```cs
if (autoPay && mesoCost > 0 && session.Currency.CanAddMeso(-mesoCost) != -mesoCost) {
    return ItemBoxError.s_err_cannot_open_multi_itembox_inventory;
}

if (!session.Item.Inventory.Consume(new[] {boxIngredient})) {
    return ItemBoxError.s_err_cannot_open_multi_itembox_inventory;
}
```

Reply:
The way it works is it's auto pay, it'll consume the boxIngredient first until needing to pay mesos for it.
So if I set it to auto pay for 10 boxes and I only have 6 boxes, it'll consume the 6 boxes first then pay for the remaining 4 in mesos. It needs to consume the ingredients first before checking mesos and paying

Reply:
I realized it's not actually paying the mesos.
I pushed a change, let me know what you think. I'm doing Math.Min because sometimes the value is -1 (aka no cost)

Reply:
Looks good, but I think it's supposed to be `!=` for mesoCost